kv/kvserver: TestCheckConsistencyInconsistent failed

kv/kvserver.TestCheckConsistencyInconsistent [failed](https://mesolite.cluster.engflow.com/invocations/default/ae8deb21-14a1-406f-9419-aa21b211ea13?testReportRun=2&testReportShard=11&testReportAttempt=1#targets-Ly9wa2cva3Yva3ZzZXJ2ZXI6a3ZzZXJ2ZXJfdGVzdA==) on release-25.4.0-rc @ [1de9c4bc217dca385a4f912dbdf828bc5629711a](https://github.com/cockroachdb/cockroach/commits/1de9c4bc217dca385a4f912dbdf828bc5629711a):


```
=== RUN   TestCheckConsistencyInconsistent
    test_log_scope.go:171: test logs captured to: outputs.zip/logTestCheckConsistencyInconsistent3348889442
    test_log_scope.go:82: use -show-logs to present logs inline
    consistency_queue_test.go:402: found a checkpoint at auxiliary/checkpoints/r5_at_29
    consistency_queue_test.go:402: found a checkpoint at auxiliary/checkpoints/r5_at_29
    consistency_queue_test.go:422: 
        	Error Trace:	pkg/kv/kvserver/consistency_queue_test.go:422
        	Error:      	Received unexpected error:
        	            	range local key /Local/Range/Table/54/1/1116473603281977345/"legacy_progress"/Transaction/"3d8fc7a6-56d2-4d4a-b108-f78c767d5639" outside of a known range
        	            	(1) assertion failure
        	            	Wraps: (2) attached stack trace
        	            	  -- stack trace:
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver/kvstorage.IterateRangeDescriptorsFromDisk
        	            	  | 	pkg/kv/kvserver/kvstorage/init.go:302
        	            	  | github.com/cockroachdb/cockroach/pkg/kv/kvserver_test.TestCheckConsistencyInconsistent
        	            	  | 	pkg/kv/kvserver/consistency_queue_test.go:422
        	            	  | testing.tRunner
        	            	  | 	GOROOT/src/testing/testing.go:1690
        	            	  | runtime.goexit
        	            	  | 	src/runtime/asm_amd64.s:1700
        	            	Wraps: (3) range local key /Local/Range/Table/54/1/1116473603281977345/"legacy_progress"/Transaction/"3d8fc7a6-56d2-4d4a-b108-f78c767d5639" outside of a known range
        	            	Error types: (1) *assert.withAssertionFailure (2) *withstack.withStack (3) *errutil.leafError
        	Test:       	TestCheckConsistencyInconsistent
    consistency_queue_test.go:291: dumping sticky VFS to /mnt/engflow/worker/work/0/exec/bazel-out/k8-fastbuild/testlogs/pkg/kv/kvserver/kvserver_test/shard_11_of_48_run_2_of_3/test.outputs/logTestCheckConsistencyInconsistent3348889442/stores
    panic.go:635: -- test log scope end --
test logs left over in: outputs.zip/logTestCheckConsistencyInconsistent3348889442
--- FAIL: TestCheckConsistencyInconsistent (48.10s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=2</code>
 - <code>shard=11</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #154145 kv/kvserver: TestCheckConsistencyInconsistent failed [A-storage C-bug C-test-failure O-robot P-2 T-storage branch-master]
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestCheckConsistencyInconsistent.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-55542

--- a/pkg/kv/kvserver/consistency_queue_test.go
+++ b/pkg/kv/kvserver/consistency_queue_test.go
@@ -418,13 +418,14 @@ func TestCheckConsistencyInconsistent(t *testing.T) {
 
 		// Find the problematic range in the storage.
 		var desc *roachpb.RangeDescriptor
-		require.NoError(t, kvstorage.IterateRangeDescriptorsFromDisk(context.Background(), cpEng,
+		require.NoError(t, kvstorage.IterateRangeDescriptorsFromCheckpoint(context.Background(), cpEng,
 			func(rd roachpb.RangeDescriptor) error {
 				if rd.RangeID == resp.Result[0].RangeID {
 					desc = &rd
 				}
 				return nil
-			}))
+			},
+		))
 		require.NotNil(t, desc)
 
 		// Compute a checksum over the content of the problematic range.

--- a/pkg/kv/kvserver/kvstorage/init.go
+++ b/pkg/kv/kvserver/kvstorage/init.go
@@ -246,6 +246,28 @@ func ReadStoreIdent(ctx context.Context, eng storage.Engine) (roachpb.StoreIdent
 func IterateRangeDescriptorsFromDisk(
 	ctx context.Context, reader storage.Reader, fn func(desc roachpb.RangeDescriptor) error,
 ) error {
+	return iterateRangeDescriptorsFromDiskHelper(ctx, reader, fn, buildutil.CrdbTestBuild)
+}
+
+// IterateRangeDescriptorsFromCheckpoint is like IterateRangeDescriptorsFromDisk
+// except it is intended to be used when reading from checkpoints. Checkpoints
+// do not correspond to full stores, and as such, certain invariant checks do
+// not hold for them.
+func IterateRangeDescriptorsFromCheckpoint(
+	ctx context.Context, reader storage.Reader, fn func(desc roachpb.RangeDescriptor) error,
+) error {
+	return iterateRangeDescriptorsFromDiskHelper(ctx, reader, fn, false /* performInvariantChecks */)
+}
+
+// iterateRangeDescriptorsFromDisk performs the actual heavy lifting of reading
+// range descriptors from the supplied storage.Reader. The caller may specify
+// whether or not invariant checks should be performed.
+func iterateRangeDescriptorsFromDiskHelper(
+	ctx context.Context,
+	reader storage.Reader,
+	fn func(desc roachpb.RangeDescriptor) error,
+	performInvariantChecks bool,
+) error {
 	log.KvExec.Info(ctx, "beginning range descriptor iteration")
 
 	// We are going to find all range descriptor keys. This code is equivalent to
@@ -296,8 +318,11 @@ func IterateRangeDescriptorsFromDisk(
 				iter.SeekGE(storage.MVCCKey{Key: keys.RangeDescriptorKey(keys.MustAddr(startKey))})
 			} else {
 				// This case shouldn't happen in practice: we have a key that isn't
-				// associated with any range descriptor.
-				if buildutil.CrdbTestBuild {
+				// associated with any range descriptor. However, when reading from
+				// checkpoints (incomplete stores), this can happen because checkpoints
+				// only guarantee to include specific key spans and may have arbitrary
+				// keys outside those spans.
+				if performInvariantChecks {
 					return errors.AssertionFailedf("range local key %s outside of a known range", key.Key)
 				}
 				iter.NextKey()

